File: data-structures/arrays/array.js
/**
 * - My custom implementation of an array data structure via a JS class
 * - For illustration purposes only, this implementation uses a JS object to store the array elements
 * - Overall space complexity: O(n), n = Length of array 
 */
class MyArray {
    /**
     * - Instantiates a new empty array
     * - Time complexity: O(1)
     * - Space complexity: O(1)
     */
    constructor() {
        this.data = {};
        this.length = 0;
    }

    /** 
     * - Returns the element at a given index
     * - Time complexity: O(1)
     * - Space complexity: O(1)
     */
    lookup(index) {
        if (index < 0 || index >= this.length) throw new Error(`Index ${index} is out of bounds.`);
        
        return this.data[index];
    }

    /** 
     * - Inserts an element at the end of the array
     * - Time complexity: O(1)
     * - Space complexity: O(1)
     */
    push(element) {
        if (element === undefined) throw new Error('This function requires 1 argument: element.');
        
        this.data[this.length] = element;
        this.length++;
        return this.length;
    }

    /** 
     * - Removes the element at the end of the array
     * - Time complexity: O(1)
     * - Space complexity: O(1)
     */
    pop() {
        if (this.length < 1) throw new Error('This array is empty.');
        
        const deletedElement = this.data[this.length - 1]; 
        delete this.data[this.length - 1];
        this.length--;
        return deletedElement;
    }

    /** 
     * - Inserts an element at a given index
     * - Time complexity: O(n), n = Length of array
     * - Space complexity: O(1)
     */
    insert(element, index) {
        if (element === undefined) {
            throw new Error('This method requires an element argument');
        } else if (index < 0 || index >= this.length) {
            throw new Error(`Index ${index} is out of bounds.`);
        } else if (index === this.length - 1) {
            return this.push(element);
        }

        // Shift each index to the left, keeping track of next element
        let currentElement = element;
        for (let i = index; i < this.length; i++) {
            let temp = this.data[i];
            this.data[i] = currentElement;
            currentElement = temp;
        }
        // Create new index for last element
        this.data[this.length] = currentElement;
        this.length++;
        return this.length;
    }
    
    /** 
     * - Removes the element at a given index
     * - Time complexity: O(n), n = Length of array
     * - Space complexity: O(1)
     */
    delete(index) {
        if (index < 0 || index >= this.length) {
            throw new Error(`Index ${index} is out of bounds.`);
        } else if (index === this.length - 1) {
            return this.pop();
        }
        
        let deletedElement = this.data[index];
        delete this.data[index]; // Remove selected element
        // Shift each index to the right
        for (let i = index; i < this.length - 1; i++) {
            this.data[i] = this.data[i + 1];
        }
        // Remove last unused index
        delete this.data[this.length - 1];
        this.length--;
        return deletedElement;
    }
}

File: data-structures/arrays/array.py
class Array:
    '''
    My custom implementation of an array data structure via a Python class
    
    For illustration purposes only, this implementation uses a Python dictionary to mimic the storage of elements in computer memory. 
    
    Overall space complexity: O(n), n = Length of array
    '''

    def __init__(self):
        '''
        Instantiates a new empty array
        
        Time complexity: O(1)
        
        Space complexity: O(1)
        '''

        self.data = {}
        self.length = 0
    
    def lookup(self, index):
        '''
        Returns the element at a given index

        Time complexity: O(1)

        Space complexity: O(1)
        '''

        assert index > 0, f'Index {index} is too small.'
        assert index < self.length, f'Index {index} is too large.'

        return self.data[index]
    
    def append(self, element):
        '''
        Inserts an element at the end of the array

        Time complexity: O(1)
        
        Space complexity: O(1)
        '''

        self.data[self.length] = element
        self.length += 1

    def pop(self):
        '''
        Removes the element at the end of the array

        Time complexity: O(1)

        Space complexity: O(1)
        '''

        assert self.length > 0, 'This list is empty.'
        del self.data[self.length - 1]
        self.length -= 1

    def insert(self, element, index):
        '''
        Inserts an element at a given index
        
        Time complexity: O(n), n = Length of array
        
        Space complexity: O(1)
        '''

        assert index > 0, f'Index {index} is too small.'
        assert index < self.length, f'Index {index} is too large.'

        if index == self.length - 1:
            return self.append(element)

        # Shift each index to the left
        # Keep track of next element in temp variable
        currentElement = element
        for i in range(index, self.length): # O(n)
            temp = self.data[i]
            self.data[i] = currentElement
            currentElement = temp
        # Create new index for last element
        self.data[self.length] = currentElement
        self.length += 1

    def delete(self, index):
        '''
        Removes the element at a given index
        
        Time complexity: O(n), n = Length of array
        
        Space complexity: O(1)
        '''

        assert index > 0, f'Index {index} is too small.'
        assert index < self.length, f'Index {index} is too large.'

        if index == self.length - 1:
            return self.pop()
        
        # Delete selected element
        del self.data[index]
        # Shift each index to the right - O(n)
        for i in range(index, self.length - 1):
            self.data[i] = self.data[i + 1]
        # Remove last unused index
        del self.data[self.length - 1]
        self.length -= 1

    def __str__(self):
        '''
        Returns string representation of array

        Time complexity: O(n)

        Space complexity: O(n)
        '''
        if self.length == 0:
            return '[]'
        
        result = '['
        for i in range(0, self.length): # O(n)
            result += str(self.data[i]) + ', '
        return result[:-2] + ']'

